daemon: expose HTTP endpoint on localhost for health checks

tklauser opened the pull request:
This is be used to replace k8s liveness/readiness probes currently
invoking `cilium status --brief`.
    
HTTP 200 corresponds to `cilium status --brief` printing `OK` and
exiting with status 0.
    
HTTP 500 corresponds to `cilium status --brief` printing an error and
exiting with status 1.

```release-note
Switch k8s liveness/readiness probes to use HTTP /healthz endpoint instead of "cilium status --brief" command.
```

gandro reviewed (changes requested):
Nice! Simpler than I would have expected.

gandro commented on daemon/cmd/healthz.go:
```diff
@@ -0,0 +1,55 @@
+// status HTTP endpoint exposed on addr. This endpoint reports the agent health
+// status and is equivalent to what the `cilium status --brief` CLI tool reports.
+func (d *Daemon) startAgentHealthzHTTPService(addr string) {
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
```
I believe you meant to use http.Handle**r**Func here and pass it to `srv` as a handler. `http.HandleFunc` will register a global handler (for use with `http.ListenAndServe`) and may panic if other code registers the `/healthz` pattern.

tklauser replied:
Oops, well spotted. I initially was using `http.ListenAndServe` and forgot to convert this to `http.HandlerFunc` when changing to `srv.ListenAndServe`.

aanm commented on install/kubernetes/quick-install.yaml:
```diff
@@ -373,11 +373,11 @@ spec:
+          httpGet:
+            host: '127.0.0.1'
+            path: /healthz
+            port: 9876
```
This probably requires documentation changes. Cilium agent is running as host network mode so this needs the node to have this port available.

joestringer replied:
Please make this port configurable for this reason (in agent and in helm templates).

joestringer replied:
Also /cc @pchaigno for impact on host firewall.

joestringer commented on install/kubernetes/cilium/charts/agent/templates/daemonset.yaml:
```diff
@@ -90,6 +90,10 @@ spec:
+        # Set the TCP port for the agent health status API. This is not the port used
+        # for cilium-health.
+        - name: CILIUM_AGENT_HEALTH_PORT
+          value: {{ .Values.agent.healthPort }}
```
Ah, I forgot that it would be an option here when providing pointers on Slack.

If this is also in the configmap (generated from the same values.yaml definition) then these lines will be unnecessary as the agent will automatically propagate a configmap entry named like `agent-health-port` into a bound env variable `CILIUM_AGENT_HEALTH_PORT` and hence into the daemon `option.AgentHealthPort` option.

aanm commented on Documentation/install/system_requirements.rst:
```diff
@@ -289,6 +289,19 @@ ICMP 8/0                 egress          ``worker-sg`` (self) health checks
+4245/tcp                 hubble relay
+6942/tcp                 operator Prometheus metrics
+9090/tcp                 cilium-agent Prometheus metrics
+9876/tcp                 cilium-agent health status API
```
It might be worth adding a new column stating this one only needs to be open on the localhost and not on all interfaces like the other ports.

gandro commented on daemon/cmd/healthz.go:
```diff
@@ -0,0 +1,89 @@
+// status and is equivalent to what the `cilium status --brief` CLI tool reports.
+func (d *Daemon) startAgentHealthHTTPService(addr string) {
+	lc := net.ListenConfig{Control: setsockoptReuseAddrAndPort}
+	ln, err := lc.Listen(context.Background(), "tcp", addr)
```
I've never used a `ListenConfig` before and I couldn't find any documentation about the context. I'm wondering however if we should we cancel this context if `srv.Serve()` returns? In the end, I guess it doesn't matter that much, since it only returns on shutdown (or error, where we exit as well).

tklauser replied:
Yeah, I guess it doesn't matter much in the end, but I could address it in a follow-up PR if need be.

gandro reviewed (approved):
The HTTP server part looks good to me.

tklauser committed:
daemon: expose HTTP endpoint on localhost for health checks

This will be used to replace liveness/readiness probes currently
invoking `cilium status --brief`.

HTTP 200 corresponds to `cilium status --brief` printing `OK` and
exiting with status 0.

HTTP 500 corresponds to `cilium status --brief` printing an error and
exiting with status 1.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
docs: document ports that should be available on each node

Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
install/kubernetes: use HTTP for agent {liveness,readiness}Probe

Use the newly introduced /healthz HTTP path exposed on localhost to
check for liveness/readiness of the Cilium agent.

In the long term this would also allow to exclude the cilium CLI tool
from the runtime image and move it into a separate image used with an
ephemeral container for debugging.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tgraf commented on daemon/cmd/agenthealth.go:
```diff
@@ -0,0 +1,89 @@
+	lc := net.ListenConfig{Control: setsockoptReuseAddrAndPort}
+	ln, err := lc.Listen(context.Background(), "tcp", addr)
+	if err != nil {
+		log.Fatal(err)
```
Please add a message string so we can grep for the error if we find it in the logs.

tgraf reviewed (changes requested):
Awesome!

joestringer commented on Documentation/install/system_requirements.rst:
```diff
@@ -289,6 +289,19 @@ ICMP 8/0                 egress          ``worker-sg`` (self) health checks
+======================== ==========  ==========================================
+Port Range / Protocol    Interface   Description
+======================== ==========  ==========================================
+4240/tcp                 pod IP      cluster health checks (``cilium-health``)
```
This port is used for the host and the cilium-health endpoint IP (note, this is *not* a pod).

tklauser replied:
With the `Interface` column omitted for now, I think this doesn't need any change.

joestringer commented on Documentation/install/system_requirements.rst:
```diff
@@ -289,6 +289,19 @@ ICMP 8/0                 egress          ``worker-sg`` (self) health checks
+4244/tcp                 pod IP      hubble server
+4245/tcp                 all         hubble relay
+6942/tcp                 pod IP      operator Prometheus metrics
+9090/tcp                 pod IP      cilium-agent Prometheus metrics
```
I'm actually not sure what `pod IP` means in this context. Cilium runs in host networking mode, and AFAIK we typically bind to the ports on all devices on the node. One possible exception would be `cilium-operator`, I didn't check whether that runs in host networking mode.

joestringer commented on Documentation/install/system_requirements.rst:
```diff
@@ -289,6 +289,19 @@ ICMP 8/0                 egress          ``worker-sg`` (self) health checks
+The following ports should also be available on each node:
+
+======================== ==========  ==========================================
+Port Range / Protocol    Interface   Description
```
I found the `interface` title here and the classification in the rows confusing. I'm not sure what I would do as a user to act upon this information.

Maybe it would be more clear to omit this column entirely?

tklauser replied:
I see your point (also your comment above) and think I'll drop this column for now.

joestringer commented on install/kubernetes/quick-install.yaml:
```diff
@@ -42,6 +42,9 @@ data:
   debug: "false"
+  # Set the TCP port for the agent health status API. This is not the port used
+  # for cilium-health.
+  agent-health-port: "9876"
```
I see we're in good company on this port ;)

https://www.speedguide.net/port.php?port=9876

On a more serious note, if this is purely used for loopback traffic then it's probably not a terrible choice since it's more likely administrators have blacklisted it on the actual network due to some of the other famous users.

tklauser replied:
Yeah, it's used for localhost-only traffic so we should be fine.

joestringer reviewed (changes requested):
Helm changes LGTM, A few more nits on the docs.

Also I didn't read anywhere a discussion of the port number but apparently this one is already used by a half dozen viruses :thinking:
